### ISSUE TITLE
large boolean array broadcast multiplication appears to have incorrect results

### ISSUE THREAD

**Issue by GordonMcGregor** [ml-explore/mlx#124]
Doing a fairly simple boolean masking.  

With a boolean array, summing ever row.  
Take the row with the highest total
create a mask from the True values in that row (invert it)
apply that mask to all the rows, using broadcasting.

```
    totals = data.sum(axis=1)              # sum all the rows 
    highest_rank = totals.argmax()         # find the biggest contributor
    highest_rank_mask=mx.logical_not(data[highest_rank])  # Create a mask for by the biggest contributing row
    data = data * highest_rank_mask               # clear the entries associated with those columns, in all rows

```

This works in PyTorch and works in numpy.

It works in MLX until the arrays get sufficiently large.  The boundary appears to be where the row x column dimension passes 2^31

Sample code attached that generates random arrays and computes the mask in both mlx and np, comparing the results.

Also checks that the mask has applied correctly - you can see the first failure at dimensions (1000x2147484)

Testcase that demonstrates the problem attached

```
size: 100*10000=1000000 ==2^(19.931568569324174)
ok: rows:100 columns:10000
size: 100*200000=20000000 ==2^(24.253496664211536)
ok: rows:100 columns:200000
size: 100*214748=21474800 ==2^(24.356141359470417)
ok: rows:100 columns:214748
size: 100*214749=21474900 ==2^(24.35614807753849)
ok: rows:100 columns:214749
size: 100*2147483=214748300 ==2^(27.678071469781486)
ok: rows:100 columns:2147483
size: 100*2147484=214748400 ==2^(27.678072141588764)
ok: rows:100 columns:2147484
size: 100*2767474=276747400 ==2^(28.04399452226751)
ok: rows:100 columns:2767474
size: 1000*10000=10000000 ==2^(23.25349666421154)
ok: rows:1000 columns:10000
size: 1000*200000=200000000 ==2^(27.5754247590989)
ok: rows:1000 columns:200000
size: 1000*214748=214748000 ==2^(27.678069454357782)
ok: rows:1000 columns:214748
size: 1000*214749=214749000 ==2^(27.678076172425854)
ok: rows:1000 columns:214749
size: 1000*2147483=2147483000 ==2^(30.99999956466885)
ok: rows:1000 columns:2147483
size: 1000*2147484=2147484000 ==2^(31.000000236476126)
MLX broadcast: Mask failed to apply
mismatch: rows:1000 columns:2147484
size: 1000*2767474=2767474000 ==2^(31.365922617154876)
MLX broadcast: Mask failed to apply
mismatch: rows:1000 columns:2767474
size: 10000*10000=100000000 ==2^(26.5754247590989)
ok: rows:10000 columns:10000
size: 10000*200000=2000000000 ==2^(30.897352853986263)
ok: rows:10000 columns:200000
size: 10000*214748=2147480000 ==2^(30.999997549245148)
ok: rows:10000 columns:214748
size: 10000*214749=2147490000 ==2^(31.000004267313212)
MLX ROW: Mask failed to apply
MLX broadcast: Mask failed to apply
mismatch: rows:10000 columns:214749
size: 10000*2147483=21474830000 ==2^(34.321927659556216)
```
[mlx_large_array_mult_issue.py.zip](https://github.com/ml-explore/mlx/files/13629564/mlx_large_array_mult_issue.py.zip)

**Comment by awni** [comment]
Great, thanks. I am able to reproduce the issue. Here is a simple test case:

```
import numpy as np
import mlx.core as mx

row = 1000
col = 2147484
mx.set_default_device(mx.cpu)
x = np.random.random( (row, col))>0.5
y = np.random.random(col) > 0.5

mlx_result = mx.array(x) * mx.array(y)
np_result = x * y

if (np.array(mlx_result) == np_result).all():
    print(f'ok: rows:{row} columns:{col}')
else:
    print(f'mismatch: rows:{row} columns:{col}')
```

This definitely seems like a bug due to overflowing int32 when indexing into the arrays. Note it works fine on the CPU so it likely is a bug in our Metal binary op kernel (possible the bug exists in other kernels as well...).

**Comment by awni** [comment]
Fix is in #125
